File: standalone-container/src/main/scala/com/yahoo/container/standalone/CloudConfigInstallVariables.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.standalone

import java.util.Optional

import com.yahoo.vespa.model.container.configserver.option.CloudConfigOptions
import com.yahoo.vespa.model.container.configserver.option.CloudConfigOptions.ConfigServer

import scala.language.implicitConversions
import scala.util.Try

/**
  * @author Tony Vaagenes
  */
class CloudConfigInstallVariables extends CloudConfigOptions {
  import CloudConfigInstallVariables._

  override val rpcPort = optionalInstallVar[Integer]("port_configserver_rpc", "services")
  override val allConfigServers = installVar("addr_configserver", "services") withDefault Array[ConfigServer]()
  override val multiTenant = optionalInstallVar[java.lang.Boolean]("multitenant")

  override val zookeeperBarrierTimeout = optionalInstallVar[java.lang.Long]("zookeeper_barrier_timeout")
  override val sessionLifeTimeSecs = optionalInstallVar[java.lang.Long]("session_lifetime")
  override val configModelPluginDirs = installVar("config_model_plugin_dirs") withDefault Array[String]()
  override val zookeeperClientPort = optionalInstallVar[Integer]("zookeeper_clientPort")
  override val zookeeperQuorumPort = optionalInstallVar[Integer]("zookeeper_quoromPort")
  override val zookeeperElectionPort = optionalInstallVar[Integer]("zookeeper_electionPort")
  override val payloadCompressionType = optionalInstallVar[java.lang.String]("payload_compression_type")
  override val environment = optionalInstallVar[java.lang.String]("environment")
  override val region = optionalInstallVar[java.lang.String]("region")
  override val system = optionalInstallVar[java.lang.String]("system")
  override val defaultFlavor = optionalInstallVar[java.lang.String]("default_flavor")
  override val defaultAdminFlavor = optionalInstallVar[java.lang.String]("default_admin_flavor")
  override val defaultContainerFlavor = optionalInstallVar[java.lang.String]("default_container_flavor")
  override val defaultContentFlavor = optionalInstallVar[java.lang.String]("default_content_flavor")
  override val useVespaVersionInRequest = optionalInstallVar[java.lang.Boolean]("use_vespa_version_in_request")
  override val hostedVespa = optionalInstallVar[java.lang.Boolean]("hosted_vespa")
  override val numParallelTenantLoaders = optionalInstallVar[java.lang.Integer]("num_parallel_tenant_loaders")
  override val dockerRegistry = optionalInstallVar[java.lang.String]("docker_registry")
  override val dockerVespaBaseImage = optionalInstallVar[java.lang.String]("docker_vespa_base_image")
  override val loadBalancerAddress = optionalInstallVar[java.lang.String]("load_balancer_address")
}

object CloudConfigInstallVariables {
  private class InstallVariable(installPkg:String, name: String) {
    val value = Environment.optionalInstallVariable(installPkg + "." + name)

    def withDefault[T](defaultValue: T)(implicit c: Converter[T]) : T = {
      value map { implicitly[Converter[T]].convert } getOrElse defaultValue
    }
  }

  private def installVar(setting:String, installPkg: String = "cloudconfig_server") = new InstallVariable(installPkg, setting)

  private def optionalInstallVar[T](setting:String, installPkg: String = "cloudconfig_server")(implicit c: Converter[T]): Optional[T] = {
    Environment.optionalInstallVariable(installPkg + "." + setting) map ( c.convert )
  }

  implicit val configServerConverter: Converter[Array[ConfigServer]] = new Converter[Array[ConfigServer]] {
    override def convert(s: String) = {
      s split "[, ]" filter { !_.isEmpty } map { toConfigServer }
    }
  }

  implicit val stringArrayConverter: Converter[Array[String]] = new Converter[Array[String]] {
    override def convert(s: String) = {
      s split "[, ]" filter { !_.isEmpty }
    }
  }

  private def toConfigServer(hostPort: String): ConfigServer = Try {
    val (host, portStr) = splitFirst(hostPort, ':')
    val port = portStr map { _.toInt }
    new ConfigServer(host, port)
  }.getOrElse(throw new IllegalArgumentException(s"Invalid config server '$hostPort'"))

  private def splitFirst(string: String, separator: Character): (String, Option[String]) = {
    val (beginning, endWithSeparator) = string span { _ != separator }
    (beginning, tailOption(endWithSeparator))
  }

  def tailOption(s: String) = {
    if (s.isEmpty) None
    else Some(s.tail)
  }

  implicit def toJavaOptional[U <% V, V](option: Option[U]): Optional[V] = option match {
    case Some(u) => Optional.of(u: V)
    case None => Optional.empty()
  }
}

File: standalone-container/src/main/java/com/yahoo/container/standalone/CloudConfigInstallVariables.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.standalone;

import com.yahoo.vespa.model.container.configserver.option.CloudConfigOptions;

import java.util.Arrays;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Stream;

/**
 * @author bjorncs
 */
public class CloudConfigInstallVariables implements CloudConfigOptions {

    @Override
    public Optional<Integer> rpcPort() {
        return Optional.ofNullable(System.getenv("VESPA_CONFIGSERVER_RPC_PORT"))
                .or(() -> getRawInstallVariable("services.port_configserver_rpc"))
                .map(Integer::parseInt);
    }

    @Override
    public Optional<Boolean> multiTenant() {
        return Optional.ofNullable(System.getenv("VESPA_CONFIGSERVER_MULTITENANT"))
                .or(() -> getInstallVariable("multitenant"))
                .map(Boolean::parseBoolean);
    }

    @Override
    public ConfigServer[] allConfigServers() {
        return Optional.ofNullable(System.getenv("VESPA_CONFIGSERVERS"))
                .or(() -> getRawInstallVariable("services.addr_configserver"))
                .map(CloudConfigInstallVariables::toConfigServers)
                .orElseGet(() -> new ConfigServer[0]);
    }

    @Override
    public int[] configServerZookeeperIds() {
        return Optional.ofNullable(System.getenv("VESPA_CONFIGSERVER_ZOOKEEPER_IDS"))
                .map(CloudConfigInstallVariables::multiValueParameterStream)
                .orElseGet(Stream::empty)
                .mapToInt(Integer::valueOf)
                .toArray();
    }

    @Override
    public Optional<Long> zookeeperBarrierTimeout() {
        return getInstallVariable("zookeeper_barrier_timeout", Long::parseLong);
    }

    @Override
    public Optional<Long> sessionLifeTimeSecs() {
        return getInstallVariable("session_lifetime", Long::parseLong);
    }

    @Override
    public String[] configModelPluginDirs() {
        return getRawInstallVariable("cloudconfig_server.config_model_plugin_dirs")
                .map(CloudConfigInstallVariables::toConfigModelsPluginDir)
                .orElseGet(() -> new String[0]);
    }

    @Override
    public Optional<Integer> zookeeperClientPort() {
        return getInstallVariable("zookeeper_clientPort", Integer::parseInt);
    }

    @Override
    public Optional<Integer> zookeeperQuorumPort() {
        return getInstallVariable("zookeeper_quorumPort", Integer::parseInt);
    }

    @Override
    public Optional<Integer> zookeeperElectionPort() {
        return getInstallVariable("zookeeper_electionPort", Integer::parseInt);
    }

    @Override
    public Optional<String> environment() {
        return Optional.ofNullable(System.getenv("VESPA_ENVIRONMENT"))
                .or(() -> getInstallVariable("environment"));
    }

    @Override
    public Optional<String> region() {
        return Optional.ofNullable(System.getenv("VESPA_REGION"))
                .or(() -> getInstallVariable("region"));
    }

    @Override
    public Optional<String> system() {
        return Optional.ofNullable(System.getenv("VESPA_SYSTEM"))
                .or(() -> getInstallVariable("system"));
    }

    @Override
    public Optional<String> cloud() {
        return Optional.ofNullable(System.getenv("VESPA_CLOUD"));
    }

    @Override
    public Optional<Boolean> useVespaVersionInRequest() {
        return getInstallVariable("use_vespa_version_in_request", Boolean::parseBoolean);
    }

    @Override
    public Optional<Boolean> hostedVespa() {
        return Optional.ofNullable(System.getenv("VESPA_CONFIGSERVER_HOSTED"))
                .or(() -> getInstallVariable("hosted_vespa"))
                .map(Boolean::parseBoolean);
    }

    @Override
    public Optional<String> loadBalancerAddress() {
        return getInstallVariable("load_balancer_address");
    }

    @Override
    public Optional<String> athenzDnsSuffix() {
        return getInstallVariable("athenz_dns_suffix");
    }

    @Override
    public Optional<String> ztsUrl() {
        return getInstallVariable("zts_url");
    }

    static ConfigServer[] toConfigServers(String configserversString) {
        return multiValueParameterStream(configserversString)
                .map(CloudConfigInstallVariables::toConfigServer)
                .toArray(ConfigServer[]::new);
    }

    static ConfigServer toConfigServer(String configserverString) {
        try {
            String[] hostPortTuple = configserverString.split(":");
            if (configserverString.contains(":")) {
                return new ConfigServer(hostPortTuple[0], Optional.of(Integer.parseInt(hostPortTuple[1])));
            } else {
                return new ConfigServer(configserverString, Optional.empty());
            }
        } catch (Exception e) {
            throw new IllegalArgumentException("Invalid config server " + configserverString, e);
        }
    }

    static String[] toConfigModelsPluginDir(String configModelsPluginDirString) {
        return multiValueParameterStream(configModelsPluginDirString).toArray(String[]::new);
    }

    private static Optional<String> getInstallVariable(String name) {
        return getInstallVariable(name, Function.identity());
    }

    private static <T> Optional<T> getInstallVariable(String name, Function<String, T> converter) {
        return getRawInstallVariable("cloudconfig_server." + name).map(converter);
    }

    private static Optional<String> getRawInstallVariable(String name) {
        return Optional.ofNullable(
                Optional.ofNullable(System.getenv(name.replace(".", "__")))
                        .orElseGet(() -> System.getProperty(name)));
    }

    private static Stream<String> multiValueParameterStream(String param) {
        return Arrays.stream(param.split("[, ]")).filter(value -> !value.isEmpty());
    }
}
